### PR TITLE
Unify usage of XMLHttpRequest and Promise

### PR DESCRIPTION
**Pull request by peppsac**
XMLHttpRequestHelper module provides a simple method to fire a
XMLHttpRequest so each driver doesn't need to implement this.

Also use Promise's reject method instead of resolve(undefined) in
case of error.

@gchoqueux r?

### CODE REVIEW

File: src/Core/Web/XMLHttpRequestHelper.js [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+define('Core/Web/XMLHttpRequestHelper', ['when'], function(when) {
+    return {
+        fetch: function(url, responseType) {
+            return when.promise(function(resolve, reject) {
+                var xhr = new XMLHttpRequest();
+                xhr.open('GET', url, true);
+                xhr.responseType = responseType;
+                xhr.crossOrigin = '';
+                xhr.onload = function(e) {
+                    resolve(this.response);
```
Comment (tbroyer) [inline src/Core/Web/XMLHttpRequestHelper.js]: This ignores and hides the response status.

File: src/Core/Web/XMLHttpRequestHelper.js [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+
+define('Core/Web/XMLHttpRequestHelper', ['when'], function(when) {
+    return {
+        fetch: function(url, responseType) {
+            return when.promise(function(resolve, reject) {
+                var xhr = new XMLHttpRequest();
+                xhr.open('GET', url, true);
+                xhr.responseType = responseType;
+                xhr.crossOrigin = '';
```
Comment (tbroyer) [inline src/Core/Web/XMLHttpRequestHelper.js]: I can't find anything wrt that property, which browser is that for? And what does it do?

File: src/Core/Web/XMLHttpRequestHelper.js [hunk 3]
```diff
@@ -0,0 +1,26 @@
+/**
+ * Generated On: 2015-10-5
+ * Class: XMLHttpRequestHelper. Simple promise-based wrapper around XMLHttpRequest
+ */
+
+
+define('Core/Web/XMLHttpRequestHelper', ['when'], function(when) {
+    return {
+        fetch: function(url, responseType) {
```
Comment (tbroyer) [inline src/Core/Web/XMLHttpRequestHelper.js]: How about using the native `fetch` when available and a polyfill otherwise? http://caniuse.com/#feat=fetch

(and similarly for promises, now's the time)

File: src/Core/Web/XMLHttpRequestHelper.js (hunk 2 above)
Comment (peppsac) [inline src/Core/Web/XMLHttpRequestHelper.js]: I dunno: it was used in every Drivers so I kept it.

I can remove it if it's useless - wdyt?

File: src/Core/Web/XMLHttpRequestHelper.js (hunk 3 above)
Comment (peppsac) [inline src/Core/Web/XMLHttpRequestHelper.js]: I have no strong feeling regarding `fetch` vs `XmlHttpRequest` and `promise` vs `when.`

But:
- for fetch: if we want to switch, this PR is a good opportunity to do it because it's focused primarly on XmlHttpRequest
- for Promise: I'd prefer to wait do it in another PR where we would switch _all_ `when` usage to Promise

File: src/Core/Web/XMLHttpRequestHelper.js (hunk 1 above)
Comment (peppsac) [inline src/Core/Web/XMLHttpRequestHelper.js]: This is from the original Drivers code.
Do you have a better proposal in mind?

File: src/Core/Web/XMLHttpRequestHelper.js (hunk 2 above)
Comment (tbroyer) [inline src/Core/Web/XMLHttpRequestHelper.js]: I suppose it originated from a _copypasta_ from the `Image` handling in `IoDriver_Image` and can be safely deleted.

File: src/Core/Web/XMLHttpRequestHelper.js (hunk 1 above)
Comment (tbroyer) [inline src/Core/Web/XMLHttpRequestHelper.js]: Maybe returning a fetch-like response; or move to fetch (using [`whatwg-fetch`](https://github.com/github/fetch))

File: src/Core/Web/XMLHttpRequestHelper.js (hunk 3 above)
Comment (tbroyer) [inline src/Core/Web/XMLHttpRequestHelper.js]: My _problem_ with this PR is that it adds a level of abstraction that's probably not really needed (standardises on `GET`, with no access to request headers et al.)

Moving to fetch, the calling code would move from:

``` js
return XMLHttpRequestHelper.fetch(url, 'arraybuffer').then(function(result) {
    return that.parseXBil(result);
});
```

to

``` js
return fetch(url).then(function(response) {
    return that.parseXbil(response.arrayBuffer());
});
```

So, for the 3 cases here, it wouldn't change much things, but is much more flexible.

As for `when` vs. native promises, `when` comes with a polyfill (https://github.com/cujojs/when/blob/master/docs/es6-promise-shim.md), and if that doesn't work, we could use `es6-promise` instead and easily adapt that to a `when` promise AFAICT:

``` js
return when.promise(fetch(url)).then(function(response) {
    return that.parseXbil(response.arrayBuffer());
});
```

Comment (peppsac) [inline src/Core/Web/XMLHttpRequestHelper.js]: Ok, `fetch` seems like the right tool for the job.
I'll update the PR to use it.

**Comment by peppsac** [comment]
I updated the branch based on @tbroyer  comments.
Let me know what you think.

File: webpack.config.js [hunk 4]
```diff
@@ -6,7 +6,7 @@ var path = require('path');
 //                                               #idcommit
 
 module.exports = {
-  entry: path.resolve(__dirname, 'src/Main.js'),
+  entry: ['whatwg-fetch', path.resolve(__dirname, 'src/Main.js')],
```
Comment (tbroyer) [inline webpack.config.js]: Not sure it's a good idea to "bundle" the polyfills; better declare them as [`externals`](https://webpack.github.io/docs/configuration.html#externals) IMO, and ask users to load them in addition to `itowns2.js` if they need it (want to support browsers with no native support).
We could revisit this later though (because the same reasoning could apply to other dependencies, such as Three.js).

File: src/Core/Commander/Providers/IoDriverXML.js [hunk 5]
```diff
@@ -19,7 +19,9 @@ define('Core/Commander/Providers/IoDriverXML', ['Core/Commander/Providers/IoDriv
     IoDriverXML.prototype.constructor = IoDriverXML;
 
     IoDriverXML.prototype.read = function(url) {
-        return XMLHttpRequestHelper.fetch(url, 'document');
+        return fetch(url).then(function(response) {
```
Comment (tbroyer) [inline src/Core/Commander/Providers/IoDriverXML.js]: Adapt the promise to a `when` promise?

File: package.json [hunk 6]
```diff
@@ -23,10 +23,11 @@
     "js-priority-queue": "^0.1.5",
     "jszip": "^2.5.0",
     "paralleljs": "^0.2.1",
-    "three": "^0.77.0",
+    "simd": "^2.0.0",
     "string_format": "^0.0.6",
-    "when": "^3.7.7",
-    "simd": "^2.0.0"
+    "three": "^0.77.0",
+    "whatwg-fetch": "^1.0.0",
```
Comment (tbroyer) [inline package.json]: We need to use a promise polyfill too. `when` comes with one but it needs to be explicitly invoked AFAICT (possibly just adding the appropriate entry to webpack's `entry`).
See also https://github.com/iTowns/itowns2/pull/95#issuecomment-230220345

Comment (peppsac) [inline package.json]: Then switch from `when` to promise polyfill should be done before this PR.
Can you take care of it @tbroyer ?

Comment (tbroyer) [inline package.json]: For reference: #100

**Comment by peppsac** [comment]
I rebased the PR on master now that #100  has landed.

@tbroyer r?

File: src/Core/Commander/Providers/IoDriverXML.js [hunk 7]
```diff
@@ -19,31 +19,9 @@ define('Core/Commander/Providers/IoDriverXML', ['Core/Commander/Providers/IoDriv
...
-
-            xhr.onerror = function() {
-
-                reject(Error("Error IoDriverXML"));
-
-            };
-
-            xhr.send(null);
+        return fetch(url).then(function(response) {
+            return response.text();
```
Comment (tbroyer) [inline src/Core/Commander/Providers/IoDriverXML.js]: Oh, just happens to me that there's no direct `fetch` equivalent to `responseType="document"`, except of course parsing the XML explicitly with `DOMParser` **and** sniffing the encoding and turning the binary response to a string.

Assuming the XML is always in UTF-8, that'd read:

``` js
var parser = new DOMParser();
return parser.parseFromString(response.text(), "application/xml");
```

(this is not strictly equivalent, besides the encoding issue, as the XHR would have parsed the response as HTML if the server had returned an XML content-type, and would have returned an `SVGDocument` if the response had an SVG content-type; but I think we can make a few assumptions and take a few shortcuts here)

Either that or keep using `XMLHttpRequest` here, and add a comment as to why it doesn't use `fetch`.

File: src/Core/Commander/Providers/IoDriver_Image.js [hunk 8]
```diff
@@ -18,32 +18,19 @@ define('Core/Commander/Providers/IoDriver_Image', ['Core/Commander/Providers/IoD
...
             }, false);
 
-
             image.addEventListener('error', function(/*event*/) {
-
-                this.src = '';
                 resolve(undefined);
-
-            }.bind(this), false);
+            }, false);
```
Comment (tbroyer) [inline src/Core/Commander/Providers/IoDriver_Image.js]: Unless the small changes (remove `progress` event handling, remove setting `src` to the empty-string) are needed (they might very-well be), maybe revert changes to this file?

(either that or put them in a specific commit: this is no longer about standardizing on XHR or promise, this is already using promise; it's possibly –probably?– broken, but that's a different issue, so it deserves a distinct commit IMO)

File: src/Core/Commander/Providers/IoDriver_JSON.js [hunk 9]
```diff
@@ -19,27 +18,11 @@ define('Core/Commander/Providers/IoDriver_JSON', ['Core/Commander/Providers/IoDr
...
-
-            xhr.onerror = function() {
-                reject(Error("Error IoDriver_JSON"));
-            };
-
-            xhr.send(null);
+        return fetch(url).then(function(response) {
+            return response.text();
+        }).then(function(text) {
+            return JSON.parse(text);
```
Comment (tbroyer) [inline src/Core/Commander/Providers/IoDriver_JSON.js]: In the first `then`, simply `return response.json();`

File: src/Core/Commander/Providers/IoDriver_XBIL.js [hunk 10]
```diff
@@ -50,35 +50,12 @@ define('Core/Commander/Providers/IoDriver_XBIL', ['Core/Commander/Providers/IoDr
...
-                resolve(undefined);
-
-                this.abort();
-            };
-
-            xhr.open("GET", url, true);
-            xhr.send(null);
-
-        }.bind(this));
+        var that = this;
```
Comment (tbroyer) [inline src/Core/Commander/Providers/IoDriver_XBIL.js]: Use `bind(this)` where needed instead

**Comment by peppsac** [comment]
Updated/rebased the branch based on @tbroyer  comments.

**Comment by tbroyer** [comment]
I see other usages of `XMLHttpRequest` in `MobileMapping/Ori`, `MobileMapping/Shader`, `PanoramicProvider` and `KMZLoader`; do you want to replace them with `fetch` now or in a followup PR?

Otherwise LGTM, feel free to merge.

**Comment by nosy-b** [comment]
I'm actually working on MobilMapping classes so i'll use fetch after finishing

**Comment by peppsac** [comment]
- updated KMZLoader
- PanoramicProvider uses XmlHttpRequest to issue a send command so I left it unmodified.
- @nosy-b will take care of MobileMapping

File: src/Renderer/ThreeExtented/KMZLoader.js [hunk 11]
```diff
@@ -28,60 +28,35 @@ define('Renderer/ThreeExtented/KMZLoader', ['Renderer/ThreeExtented/jszip.min',
...
-
-                    var zip = new JSZip(this.response);
-                    var collada = undefined;
-                    var coordCarto = undefined;
-                    for (var name in zip.files) {
-
-                        if (name.toLowerCase().substr(-4) === '.dae') {
-                            collada = scopeLoader.parse(zip.file(name).asText());
-                        } else if (name.toLowerCase().substr(-4) === '.kml') {
+        KMZLoader.prototype.parsCollada = function(buffer, scopeLoader) {
```
Comment (tbroyer) [inline src/Renderer/ThreeExtented/KMZLoader.js]: Typo: should be `parseCollada` (missing `e`)

File: src/Renderer/ThreeExtented/KMZLoader.js [hunk 12]
```diff
@@ -28,60 +28,35 @@ define('Renderer/ThreeExtented/KMZLoader', ['Renderer/ThreeExtented/jszip.min',
...
-                };
-
-                xhr.send(null);
+            collada.coorCarto = coordCarto;
+            return collada;
+        }
 
+        KMZLoader.prototype.load = function(url) {
+            return fetch(url).then(function(response) {
+                return this.parseCollada(response.arrayBuffer(), this.colladaLoader);
```
Comment (tbroyer) [inline src/Renderer/ThreeExtented/KMZLoader.js]: `response.arrayBuffer()` returns a `Promise`, so it should probably be "cascaded" as in `IoDriver_XBil`.

Not sure we need to pass the `ColladaLoader` as argument (AFAICT, `scopeLoader` was used much like `var that = this;` as the callbacks didn't use `bind(this)`)

File: src/Renderer/ThreeExtented/KMZLoader.js (hunk 11 above)
Comment (peppsac) [inline src/Renderer/ThreeExtented/KMZLoader.js]: oops :)

**Comment by peppsac** [comment]
Updated again: should be typo-free now...

**Comment by tbroyer** [comment]
LGTM
